osu.Server.Spectator/Entities/ConnectionState.cs

```diff
@@ -42,5 +58,36 @@ public ConnectionState(HubLifetimeContext context)
         /// <param name="context">The hub context to retrieve information from.</param>
         public void RegisterConnectionId(HubLifetimeContext context)
             => ConnectionIds[context.Hub.GetType()] = context.Context.ConnectionId;
+
+        private bool tryGetClientSessionID(HubLifetimeContext context, out Guid clientSessionId)
+        {
+            clientSessionId = Guid.Empty;
+            return context.Context.GetHttpContext()?.Request.Headers.TryGetValue(HubClientConnector.CLIENT_SESSION_ID_HEADER, out var value) == true
+                   && Guid.TryParse(value, out clientSessionId);
+        }
+
+        public bool IsConnectionFromSameClient(HubLifetimeContext context)
+        {
+            if (tryGetClientSessionID(context, out var clientSessionId))
+                return ClientSessionId == clientSessionId;
+
+            return TokenId == context.Context.GetTokenId();
+        }
+
+        public bool IsInvocationPermitted(HubInvocationContext context)
+        {
+            bool hubRegistered = ConnectionIds.TryGetValue(context.Hub.GetType(), out string? registeredConnectionId);
+            bool connectionIdMatches = registeredConnectionId == context.Context.ConnectionId;
+
+            return hubRegistered && connectionIdMatches;
+        }
+
+        public bool CanCleanUpConnection(HubLifetimeContext context)
+        {
+            bool hubRegistered = ConnectionIds.TryGetValue(context.Hub.GetType(), out string? registeredConnectionId);
+            bool connectionIdMatches = registeredConnectionId == context.Context.ConnectionId;
+
+            return hubRegistered && connectionIdMatches;
+        }
```

Comment:
Readers will note that the content of this method is mostly moved from `ConcurrentConnectionLimiter`, except for the fact that the `tokenIdMatches` check is gone.

As far as I can tell that check was mostly dead / bogus anyway; the signalr communication protocol is such that the JWT is only sent at the start of a connection, after which a protocol switch to webscokets happens, and headers are not sent anymore. Which, in turn, means that the JWT is not being sent over the wire, and the previous check would just keep checking the same stored JWT from the initial connection, probably fetching it from a dictionary keyed by connection ID or something. Thus, the connection ID should be more than enough to enforce concurrency at this point.